File: verifiers/node/src/crl.ts
import { readFileSync } from "node:fs";

function normalize(serialHex: string): string {
  const s = serialHex.trim().toLowerCase().replace(/^0x/, "").replace(/^0+/, "");
  return s || "0";
}

export class AttestationCrl {
  constructor(private revoked: Set<string>) {}
  isRevoked(cert: X509Certificate): boolean {
    return this._revoked.has(normalize(cert.serialNumber.toString(16)));
  }
  firstRevoked(...chains: X509Certificate[][]): string | null {
    for (const chain of chains) for (const cert of chain) {
      const h = normalize(cert.serialNumber.toString(16));
      if (this._revoked.has(h)) return h;
    }
    return null;
  }
  get size(): number { return this._revoked.size; }
  static parse(text: string): AttestationCrl {
    const set = new Set<string>();
    for (const line of text.split("\n")) {
      const s = normalize(line.split("#")[0].trim());
      if (s) set.add(s);
    }
    return new AttestationCrl(set);
  }
  static fromFile(p: string): AttestationCrl {
    return AttestationCrl.parse(readFileSync(p, "utf8"));
  }
}

File: verifiers/python/src/deviceintelligence_verifier/crl.py
"""Offline keybox-revocation list (AttestationCrl.kt port)."""
import os

def _normalize(serial_hex: str) -> str:
    s = serial_hex.strip().lower().removeprefix("0x").lstrip("0")
    return s or "0"

class AttestationCrl:
    def __init__(self, revoked: set):
        self._revoked = revoked

    def is_revoked(self, cert) -> bool:
        return _normalize(format(cert.serial_number, "x")) in self._revoked

    def first_revoked(self, *chains) -> str | None:
        for chain in chains:
            for cert in chain:
                h = _normalize(format(cert.serial_number, "x"))
                if h in self._revoked:
                    return h
        return None

    @property
    def size(self) -> int:
        return len(self._revoked)

    @staticmethod
    def parse(text: str) -> "AttestationCrl":
        serials = {_normalize(ln.split("#")[0].strip()) for ln in text.splitlines()}
        serials.discard("")
        return AttestationCrl(serials)

    @staticmethod
    def default() -> "AttestationCrl":
        p = os.path.join(os.path.dirname(__file__), "resources", "attestation-crl.txt")
        return AttestationCrl.parse(open(p, encoding="utf-8").read())
